<div style="border:solid green 2px; padding: 20px">
<b>Привет, Алексей!</b>

Меня зовут Александр Пономаренко, и я буду проверять твой проект. Предлагаю общаться на «ты» :) Но если это не удобно - дай знать, и мы перейдем на "вы". 

Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе стать data science. Ты уже проделал большую работу над проектом, но давай сделаем его еще лучше. Ниже ты найдешь мои комментарии - **пожалуйста, не перемещай, не изменяй и не удаляй их**. Увидев у тебя ошибку, в первый раз я лишь укажу на ее наличие и дам тебе возможность самой найти и исправить ее. На реальной работе твой начальник будет поступать так же, а я пытаюсь подготовить тебя именно к работе аналитиком. Но если ты пока не справишься с такой задачей - при следующей проверке я дам более точную подсказку. Я буду использовать цветовую разметку:

<div class="alert alert-danger">
<b>Комментарий ревьюера ❌:</b> Так выделены самые важные замечания. Без их отработки проект не будет принят. </div>

<div class="alert alert-warning">
<b>Комментарий ревьюера ⚠️:</b> Так выделены небольшие замечания. Я надеюсь, что их ты тоже учтешь - твой проект от этого станет только лучше. Но настаивать на их отработке не буду.

</div>

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> Так я выделяю все остальные комментарии.</div>

Давай работать над проектом в диалоге: **если ты что-то меняешь в проекте или отвечаешь на мои комменатри — пиши об этом.** Мне будет легче отследить изменения, если ты выделишь свои комментарии:
<div class="alert alert-info"> <b>Комментарий студента:</b> Например, вот так.</div>

Всё это поможет выполнить повторную проверку твоего проекта оперативнее. 

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка</a></span></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Выводы</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Проект для «Викишоп» с BERT

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 
Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.
Постройте модель со значением метрики качества F1 не меньше 0.75. 

In [ ]:
!pip install transformers

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
!pip install jupyter_server



Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
!pip install catboost

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


## Импорт библиотек

In [ ]:
import numpy as np
import pandas as pd
import torch
import transformers
from tqdm import notebook
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from transformers import AutoTokenizer, DistilBertTokenizer, DistilBertConfig, DistilBertForSequenceClassification
import warnings
warnings.filterwarnings("ignore")
import requests
import os
from sklearn.metrics import f1_score
from torch.utils.data import TensorDataset
from catboost import CatBoostClassifier
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import StratifiedKFold
from sklearn.model_selection import GridSearchCV
from torch.utils.data import Dataset, DataLoader, TensorDataset
from transformers import DistilBertTokenizer, DistilBertForSequenceClassification, PreTrainedModel
from transformers import AutoTokenizer, AutoModelForSequenceClassification

import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer
from sklearn.feature_extraction.text import TfidfVectorizer
from torch.nn.utils.rnn import pad_sequence
from tqdm.notebook import tqdm

nltk.download('punkt')

# загрузка стоп-слов на английском языке
nltk.download('stopwords')
stopwords = set(stopwords.words('english'))

# загрузка модуля лемматизации слов из библиотеки nltk
nltk.download('wordnet')
lemmatizer = WordNetLemmatizer()
from sklearn.pipeline import Pipeline


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


In [ ]:
df = pd.read_csv('https://code.s3.yandex.net/datasets/toxic_comments.csv', index_col=[0])


<div class="alert alert-warning">
<b>Комментарий ревьюера ⚠️:</b>Да, хорошо, только лучше, чтобы и на платформе ЯП запускался код
    
    
```python
path_yandex = 'datasets/toxic_comments.csv'    
path = 'toxic_comments.csv'

if os.path.exists(path):
    data = pd.read_csv(path, index_col=[0])
    display(data.head(5))
elif os.path.exists(path_yandex):
    data = pd.read_csv(path_yandex, index_col=[0])
    display(data.head(5))
else:
    print('Ошибка в считывании данных')   
```

</div>

<div class="alert alert-info"> <b>Комментарий студента:</b> Поправил </div>


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 159292 entries, 0 to 159450
Data columns (total 2 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   text    159292 non-null  object
 1   toxic   159292 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 3.6+ MB


In [ ]:
df.isna().mean()

text     0.0
toxic    0.0
dtype: float64

## Подготовим выборки

In [ ]:
df['toxic'].unique()

array([0, 1])

In [ ]:
# функция для предобработки текста
def lemmatize_text(text):
    # токенизация
    tokens = word_tokenize(text.lower())
    
    # удаление стоп-слов и неалфавитных символов
    tokens = [t for t in tokens if t not in stopwords and t.isalpha()]
    
    # лемматизация
    lemmatized_tokens = [lemmatizer.lemmatize(t) for t in tokens]
    
    # объединение слов в строку и возврат
    return " ".join(lemmatized_tokens)

# предобработка текста
df['text'] = df['text'].apply(lemmatize_text)


In [ ]:
features = df.drop(columns=['toxic'], axis=1)
target = df['toxic']
features.shape, target.shape

((159292, 1), (159292,))

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(features['text'], target, test_size = 0.2, random_state=45) 
X_train.shape[0]/df.shape[0]

0.7999962333324963

In [ ]:
# Вычисление весов классов
def class_weights(y):
    class_weights = {}
    n_classes = len(np.unique(y))
    bincount = np.bincount(y)
    n_samples = len(y)
    for i in range(n_classes):
        class_weights[i] = n_samples / (n_classes * bincount[i])
    return print("Weights:", class_weights,
                 "relation: {:.2f}".format(class_weights[1]/class_weights[0]))

In [ ]:
class_weights(y_train)

Weights: {0: 0.5566023725911561, 1: 4.916775985801373} relation: 8.83


In [ ]:
class_weights(y_test)

Weights: {0: 0.5563530315730651, 1: 4.936318562132011} relation: 8.87


Разделили выборки в соотношении 0.8:0.2: X_train:X_test. Вес класса в соотношении 1 к 8.83 для обучающей и 8.87 для тестовой

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> Хорошо</div>

<div class="alert alert-danger">
<b>Комментарий ревьюера ❌:</b> Перед tfidf следует данные обработать: удалить мусор + привести слова в нормальную форму (провести лемматизицаю)
    
+  https://webdevblog.ru/podhody-lemmatizacii-s-primerami-v-python/

</div>

In [ ]:
df['text'].head(10)

0    explanation edits made username hardcore metal...
1    match background colour seemingly stuck thanks...
2    hey man really trying edit war guy constantly ...
3    ca make real suggestion improvement wondered s...
4                        sir hero chance remember page
5               congratulation well use tool well talk
6                          cocksucker piss around work
7    vandalism matt shirvington article reverted pl...
8    sorry word offensive anyway intending write an...
9                 alignment subject contrary dulithgow
Name: text, dtype: object

<div class="alert alert-info"> <b>Комментарий студента:</b> Добавил лемматизицаю перед разделением на выборки </div>


## Catboost

In [ ]:
%%time

count_tf_idf = TfidfVectorizer()

# преобразование TF-IDF для тестовой выборки
X_train_tfidf = count_tf_idf.fit_transform(X_train)

# преобразование TF-IDF для тестовой выборки
X_test_tfidf = count_tf_idf.transform(X_test)

# вывод размеров матриц TF-IDF
print("Размер X_train_tfidf:", X_train_tfidf.shape)
print("Размер X_test_tfidf:", X_test_tfidf.shape)


Размер X_train_tfidf: (127433, 129078)
Размер X_test_tfidf: (31859, 129078)
CPU times: user 5.36 s, sys: 97.7 ms, total: 5.46 s
Wall time: 5.53 s


<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> Отлично, молодец, верно используешь Tfidf.
    
    
Совет: Внутри кросс-валидации происходит разбиение выборки на train и valid. Однако, в таком случае векторизатор обучен на всей выборке(train), а это не совсем корректно. Чтобы избежать это можно воспользоваться Pipeline:
    
```python
pipeline = Pipeline([
    ('tfidf', TfidfVectorizer(stop_words=stopwords)),
    ('logreg', LogisticRegression(random_state=42)),
])
parameters = {
    'tfidf__max_df': (0.25, 0.5, 0.75),
    'tfidf__ngram_range': [(1, 1), (1, 2), (1, 3)],
    'logreg__C': [1,2,6]
}

grid_search_tune = RandomizedSearchCV(pipeline, parameters, cv=3, n_jobs=-1, scoring='f1', verbose=3)
grid_search_tune.fit(train_features, train_targets)
  
    
```
    
Это просто каркас, можешь сам выбрать какие параметры использовать для подбора:) 
    
+  https://runebook.dev/ru/docs/scikit_learn/modules/generated/sklearn.model_selection.halvinggridsearchcv - тут про HalvingGridSearchCV
    
+  https://www.rupython.com/python-sklearn-pipeline-pipeline-28301.html - про pipeline
+  https://towardsdatascience.com/how-to-use-sklearn-pipelines-for-ridiculously-neat-code-a61ab66ca90d

</div>

In [ ]:
%%time
pipeline = Pipeline([
    ('tfidf', TfidfVectorizer()),
    ('catboost', CatBoostClassifier(iterations=500, random_seed=42, loss_function='Logloss', task_type='GPU', verbose=100))
])

parameters = {
    'tfidf__max_df': (0.25, 0.5),
    'tfidf__ngram_range': [(1, 1), (1, 2)],
    'catboost__learning_rate': [0.03, 0.1],
    'catboost__depth': [6, 10],
    'catboost__l2_leaf_reg': [1, 3],
    'catboost__class_weights': [[1, 8.8], [1, 8.9]]
}

cat_grid_search = GridSearchCV(pipeline, parameters, cv=3, n_jobs=1, scoring='f1')
cat_grid_search.fit(X_train, y_train)
cat_grid_search.best_estimator_


0:	learn: 0.6777453	total: 251ms	remaining: 2m 5s
100:	learn: 0.4459782	total: 14.3s	remaining: 56.4s
200:	learn: 0.3975185	total: 26.7s	remaining: 39.7s
300:	learn: 0.3723979	total: 38.6s	remaining: 25.5s
400:	learn: 0.3525194	total: 49.9s	remaining: 12.3s
499:	learn: 0.3370244	total: 1m 3s	remaining: 0us
0:	learn: 0.6780828	total: 266ms	remaining: 2m 12s
100:	learn: 0.4478262	total: 14.9s	remaining: 58.8s
200:	learn: 0.3991734	total: 27.9s	remaining: 41.5s
300:	learn: 0.3737235	total: 40.3s	remaining: 26.6s
400:	learn: 0.3544239	total: 53.2s	remaining: 13.1s
499:	learn: 0.3390798	total: 1m 5s	remaining: 0us
0:	learn: 0.6775478	total: 227ms	remaining: 1m 53s
100:	learn: 0.4473966	total: 15s	remaining: 59.4s
200:	learn: 0.4004797	total: 28.1s	remaining: 41.8s
300:	learn: 0.3752910	total: 40.7s	remaining: 26.9s
400:	learn: 0.3550584	total: 52.9s	remaining: 13.1s
499:	learn: 0.3392314	total: 1m 4s	remaining: 0us


In [ ]:
print(f"Best params: {cat_grid_search.best_params_}")
print(f'RMSE модели на базе библиотеки Catboost {-cat_grid_search.best_score_}')
cat_rmse = -(cat_grid_search.best_score_)

<div class="alert alert-info"><b>Комментарий ревьюера ✔️:</b> Попробовал с pipilin в Colab, правда google после трех прогонов падает, думаю только платная версия будет правильно работать. Правда надо ставить n_jobs = 1. Пока дошел до этого  -- сколько нервов </div>


In [ ]:
#%%time

#model = CatBoostClassifier(loss_function='Logloss', eval_metric='F1')

#model.fit(X_train_tfidf, y_train, eval_set=(X_test_tfidf, y_test), verbose=100, early_stopping_rounds=100)


#y_pred = model.predict(X_test_tfidf)


#f1 = f1_score(y_test, y_pred)
#print("F1 Score:", f1)

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> Хорошо:)
    
    
P.S что такое 1+1?)
</div>

<div class="alert alert-info"><b> Осталось со старого проекта  </div>


## Bert

Стоит отметить, что модель обучена для работы с предложениями до 512 символов. 

In [ ]:
%%time
1+1
# Использование токенизатора и модели `toxic-bert`
tokenizer = AutoTokenizer.from_pretrained('unitary/toxic-bert')
model = AutoModelForSequenceClassification.from_pretrained('unitary/toxic-bert')


<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> Молодец, что взял токсичный берт
</div>

In [ ]:
%%time
1+1
# код для подготовки данных
max_len = 128

tokenized = df['text'].apply(lambda x: tokenizer.encode(x, add_special_tokens=True, max_length=max_len, truncation=True))
padded = np.array([i + [0]*(max_len - len(i)) for i in tokenized.values])
attention_mask = np.where(padded != 0, 1, 0)

input_ids = torch.tensor(padded)  
attention_mask = torch.tensor(attention_mask)

dataset = TensorDataset(input_ids, attention_mask, torch.tensor(df['toxic'].values))

In [ ]:
%%time
1+1
# Подготовка DataLoader
batch_size = 32
data_loader = DataLoader(dataset, batch_size=batch_size)

In [ ]:
%%time
1+1
# Выполнение прогнозирования
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model.to(device)
model.eval()
device

In [ ]:
predictions = []

for batch in tqdm(data_loader):
    input_ids, attention_mask, _ = [b.to(device) for b in batch]

    with torch.no_grad(): 
        batch_predictions = model(input_ids, attention_mask=attention_mask)

    predictions.append(batch_predictions.logits[:, 0].cpu().numpy())

features = np.concatenate(predictions, axis=0)


In [ ]:
# Apply threshold
threshold = 0.5
class_labels = (predictions_array > threshold).astype(int)

# Calculate the F1 score
f1 = f1_score(df['toxic'], class_labels, average='weighted')
print(f"F1 score: {f1}")


<div class="alert alert-info"><b> Поменял на графическийб использовал, чуть подправил для сравнения с df['toxic']  </div>


<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> Дааа, все верно:)

Давай воспользуемся `Google Colab` с `GPU`, так получится обработать все данные +- за 5-20 минут 
    
Сначала настроем среду:
    
+  `Среда выполнения` -> `сменить среду выполнения`    
+  Теперь выбираем `GPU` и `сохранить`    
    
Объявляем `BERT`:
    
```python
tokenizer = BertTokenizer.from_pretrained("unitary/toxic-bert")
model = BertModel.from_pretrained("unitary/toxic-bert")    
```

Далее создаем padded и attention_mask (максимальную длину возьмем 128 этого вполне должно хватить)
    
```python
from torch.nn.utils.rnn import pad_sequence

tokenized = df['text'].progress_apply(lambda x: tokenizer.encode(x, max_length=128, truncation=True, add_special_tokens=True)) #обрежет под нужное кол-во токенов

padded = pad_sequence([torch.as_tensor(seq) for seq in tokenized], batch_first=True) #добьет нулями  

attention_mask = padded > 0
attention_mask = attention_mask.type(torch.LongTensor)    
```
    
Далее нам нужно создать загрузчик и датасет
    
```python
from torch.utils.data import TensorDataset, DataLoader
dataset = TensorDataset(attention_mask, padded)
dataloader = DataLoader(dataset, batch_size=32, shuffle=False, num_workers=0) # создает загрузчик данных с батчом 32

device = torch.device("cuda:0") if torch.cuda.is_available() else torch.device("cpu") #проверяет доступно ли нам GPU или нет
print(f'Device: {device}')
```
Про `DataLoader` и `TensorDataset`. Огромный плюс, в том, что тебе не придется самому делать батчи + удобно очень использовать и выгружать данные (на больших проектах очень выручают, поэтому будет круто если разберешься, там нет ничего сложного, если что могу помочь, если не получится) P.S если что, не обращай внимание на collate_fn
                
+  https://towardsdatascience.com/how-to-use-datasets-and-dataloader-in-pytorch-for-custom-text-data-270eed7f7c00    
    
 
Далее мы, чутка переписываем получения embeddings. Тут мы просто переводим батчи и модель на GPU в процессе выполнения, где и происходят все вычисления
    
```python
embeddings = []
model.to(device)
model.eval() 
for attention_mask, padded in notebook.tqdm(dataloader):
    attention_mask, padded = attention_mask.to(device), padded.to(device)
    
    with torch.no_grad(): 
        batch_embeddings = model(padded, attention_mask=attention_mask)
        
    embeddings.append(batch_embeddings[0][:,0,:].cpu().numpy()) 

features = np.concatenate(embeddings)       
```
    
P.S У меня пример для получения embedding, но у тебя будет аналогично:)
</div>



## Вывод

Мы построили инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 
Обучили модель классифицировать комментарии на позитивные и негативные.
Построил модель со значением метрики качества F1 = 0.752(обучалась и предсказывала 30 минут) для catboost и Взяли обученную модель Bert-toxic и построили на ней прогноз и измерили F1= 0.85(предсказывала 5 часов 30 минут)
Использовали 2 способа построения определение токсичных коментарий:
1. нейроную сеть bert-toxic
2. модель машинного обучения catboost
У обоих моделей есть плюсы и минусы.

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> 

+  Сделай предобработку для tfidf
+  Попробуй прогнать модельку на GPU (если нет времени, то главное проведи предобработку)
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> 

Для работы с текстами используют и другие подходы. Например, сейчас активно используются RNN (LSTM) и трансформеры (BERT и другие с улицы Сезам, например, ELMO). НО! Они не являются панацеей, не всегда они нужны, так как и TF-IDF или Word2Vec + модели из классического ML тоже могут справляться. \
BERT тяжелый, существует много его вариаций для разных задач, есть готовые модели, есть надстройки над библиотекой transformers. Если, обучать BERT на GPU (можно в Google Colab или Kaggle), то должно быть побыстрее.\
https://huggingface.co/transformers/model_doc/bert.html \
https://t.me/renat_alimbekov \
https://colah.github.io/posts/2015-08-Understanding-LSTMs/ - Про LSTM \
https://web.stanford.edu/~jurafsky/slp3/10.pdf - про энкодер-декодер модели, этеншены\
https://pytorch.org/tutorials/beginner/transformer_tutorial.html - официальный гайд
по трансформеру от создателей pytorch\
https://transformer.huggingface.co/ - поболтать с трансформером \
Библиотеки: allennlp, fairseq, transformers, tensorflow-text — множество реализованных
методов для трансформеров методов NLP \
Word2Vec https://radimrehurek.com/gensim/models/word2vec.html 
    
</div>

## Чек-лист проверки

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Данные загружены и подготовлены
- [x]  Модели обучены
- [x]  Значение метрики *F1* не меньше 0.75
- [x]  Выводы написаны

<font color='blue'><b>Итоговый комментарий ревьюера</b></font>
<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>Алексей, получился хороший проект! 
    
Если есть  если есть какие либо вопросы я с удовольствием на них отвечу:) <br> Исправь, пожалуйста, замечания и жду проект на следующую проверку:) </div>
